woocommerce/rest-api/Debug_Controller.php

```diff
@@ -0,0 +1,256 @@
+ * @license   http://www.gnu.org/licenses/gpl-3.0.html GNU General Public License v3.0
+ */
+
+namespace SkyVerge\WooCommerce\PluginFramework\v5_5_0\REST_API;
+
+use SkyVerge\WooCommerce\PluginFramework\v5_5_0\SV_WC_Payment_Gateway_Plugin;
+use SkyVerge\WooCommerce\PluginFramework\v5_5_0\SV_WC_Plugin;
+
+defined( 'ABSPATH' ) or exit;
+
+if ( ! class_exists( '\\SkyVerge\\WooCommerce\\PluginFramework\\v5_5_0\\REST_API\\Debug_Controller' ) ) :
+
+
+/**
+ * The plugin REST API Debug endpoint.
+ *
+ * @since 5.5.0-dev
+ */
+abstract class Debug_Controller extends \WC_REST_Controller {
+
+
+	/** @var SV_WC_Plugin main instance */
+	protected $plugin;
+
+	/** @var string endpoint namespace */
+	protected $namespace;
+
+	/** @var string the route base */
+	protected $rest_base;
+
+
+	/**
+	 * Debug controller constructor.
+	 *
+	 * @since 5.5.0-dev
+	 *
+	 * @param SV_WC_Plugin $plugin main instance
+	 */
+	public function __construct( SV_WC_Plugin $plugin ) {
+
+		$this->plugin    = $plugin;
+		$this->namespace = 'wc/v1';
+		$this->rest_base = $plugin->get_id();
+	}
+
+
+	/**
+	 * Registers the routes for the plugin debug endpoint.
+	 *
+	 * @since 5.5.0-dev
+	 */
+	public function register_routes() {
+
+		// endpoint: 'wc/v<n>/<plugin_id>/debug/'
+		register_rest_route( $this->namespace, "/{$this->rest_base}/debug", [
+			// GET the debug mode status
+			[
+				'methods'             => \WP_REST_Server::READABLE,
+				/* @see Debug_Controller::get_item() */
+				'callback'            => [ $this, 'get_item' ],
+				/* @see Debug_Controller::get_item_permissions_check() */
+				'permission_callback' => [ $this, 'get_item_permissions_check' ],
+			],
+			// UPDATE (toggle) the debug mode
+			[
+				'methods'             => \WP_REST_Server::EDITABLE,
+				/** @see Debug_Controller::update_item() */
+				'callback'            => [ $this, 'update_item' ],
+				/** @see Debug_Controller::update_item_permissions_check() */
+				'permission_callback' => [ $this, 'update_item_permissions_check' ],
+			],
+		], true );
+	}
+
+
+	/**
+	 * Determines if a given request has access to read the plugin debug mode status.
+	 *
+	 * @since 5.5.0-dev
+	 *
+	 * @param \WP_REST_Request $request request object
+	 * @return true|\WP_Error
+	 */
+	public function get_item_permissions_check( $request ) {
+
+		if ( ! wc_rest_check_manager_permissions( 'settings', 'read' ) ) {
+			return new \WP_Error( 'wc_' . $this->get_plugin()->get_id() . '_rest_cannot_view', __( 'Sorry, you cannot list resources.', 'wc-plugin-framework' ), [ 'status' => rest_authorization_required_code() ] );
+		}
+
+		return true;
+	}
+
+
+	/**
+	 * Gets the plugin debug mode status.
+	 *
+	 * @since 5.5.0-dev.1
+	 *
+	 * @param \WP_REST_Request $request API request
+	 * @return \WP_Error|\WP_REST_Response API response
+	 */
+	public function get_item( $request ) {
+
+		$plugin     = $this->get_plugin();
+		$plugin_id  = $plugin->get_id();
+		$debug_mode = [
+			'plugin' => [
```

Comment:
any need to wrap this in `plugin`? We're already requesting a specific plugin via the route, so we know the response will contain the plugin's data. `gateways` can still be the sub object, but I don't think we'd need to separate this out.

Reply:
My rationale was that there could be debug modes and logs for different entities, and to keep everything in key-value paris for each debug component for consistency. 

What if a non-gateway plugin has a debug mode for itself plus a third part API? Seems more consistent to me, to expect that the debug data for the plugin specifically is in the key `plugin`; then `gateways` and so on. 

This makes it also more explicit when actors need to call each identity, you can see what debug mode they are touching. Being a sensitive area, I wanted the minimum ambiguity.  

Perhaps one could add a new key (not sure of the naming -- `components`?) that lists the available debug components... Like `plugin`, `gateways`... so a third party knows these keys exist?

Reply:
@unfulvio thanks for more details! I understand what you mean about the different component considerations.

I think breaking things out further is overly complicated for what the original story is trying to achieve - allow third-party apps to toggle debug logging.

I still find the response structure and data to be a bit confusing as-is. Taking a plugin like Memberships as an example (to get away from gateways) - the plugin has no debug mode. As such, I wouldn't expect the response from this endpoint to show the debug mode as `log`, since that isn't really true. The base plugin `->log()` method is purely internal, and doesn't represent any sort of debug state. Now, we could add a debug mode to Memberships in the future, and it wouldn't be a bad idea, but right now there is no such concept.

Similarly, a gateway plugin like Authorize.Net doesn't have a debug more. Its gateways do, but the plugin itself does not.

With that in mind, it seems like we should return something that more accurately represents the plugin's lack of debug mode, like `null`. Also, is there any reason to include the plugin ID again inside `plugin`, when we're already requesting that plugin ID? It seems like that would make sense if we were listing a bunch of plugins and their debug modes, but this is data for a specific plugin.

Considering a plugin like AvaTax - that plugin _does_ have a debug mode since the API debugging is no separated from the plugin itself.

Reply:
hey @ChaseWiseman fair points, I've updated the response to omit the `plugin` key; additionally, I'm listing the available debug modes in a `debug_modes` key; this matches a PHP method that can also be used to validate modes being set or when listing modes in settings for example - I've updated MailChimp Sync test release with the new version; Authorize.Net didn't require anything, just need to update the FW from composer